### ISSUE TITLE
Aircraft may sit idle in RA missions

### ISSUE THREAD

**Issue by cjshmyr** [OpenRA/OpenRA#18261]
Mentioned in https://github.com/OpenRA/OpenRA/issues/4989#issuecomment-633302909

Aircraft in RA missions may exhibit this behavior
- Pick a random target
- Attack the target
- Go re-arm, but if our target is still alive, sit on the airfield indefinitely (or until we're kicked off, then circle indefinitely).

Several RA missions use a [`TargetAndAttack` function](https://github.com/OpenRA/OpenRA/search?l=Lua&q=TargetAndAttack) for aircraft logic. As a proof of concept, this snippet below could be used in-place of the existing function to fix the problem. The only difference is the target will not persist after re-arming, and a new target is chosen (would be simple to support this).

```Lua
HuntWithAircraft = function(aircraft)
	Trigger.OnIdle(aircraft, function()
		Media.DisplayMessage(tostring(aircraft) .. ' idle triggered')

		local target = nil
		local enemies = Utils.Where(Map.ActorsInWorld, function(self) 
			return self.Owner == player and aircraft.CanTarget(self) 
		end)
		if #enemies > 0 then
			target = Utils.Random(enemies)
		end

		if target ~= nil then
			Media.DisplayMessage(tostring(aircraft) .. ' attacking target ' .. tostring(target))
			aircraft.Attack(target)
		else
			Media.DisplayMessage(tostring(aircraft) .. ' no targets found, returning to base.')
			aircraft.ReturnToBase()
		end	
	end)
end
```

I'd like to test this a bit longer, then submit a pull request to move this into a `campaign-global.lua` like we've done for TD and D2K.

If it's worth doing, aircraft in the original campaign also seem to prioritize certain actors (e.g. defenses in soviet09), and we could expand this to support it.

Interested in feedback from any mission maintainers.

**Comment by deleted-user-1** [comment]
I noticed something similar with `NavalAI` (which is a camouflaged air force general). Problem might not be limited to missions.

**Comment by abcdefg30** [comment]
Did we fix aircraft not triggering the idle trigger? I don't remember that and am not sure if that is something to fix on the Lua side (perhaps by removing the `ReturnToBase`). Imho persisting targets is the better behaviour as well.

**Comment by cjshmyr** [comment]
I wasn't sure what the original issue was with aircraft and idle triggers. Idle seems to trigger if we've killed a target and have ammo remaining, or after we've landed and fully re-armed. Agreed on persisting targets.

**Comment by abcdefg30** [comment]
The idle trigger never fires when the aircraft is flying in circles, that is at least one of the problems. The `Wait` activity was meant to prevent that (see the comment), so something else in the aircraft code has regressed and so it remains stuck in one of the previous activities.

**Comment by Smittytron** [comment]
> The idle trigger never fires when the aircraft is flying in circles

Stationary helicopters have the same issue with re-engaging. Also` IdleBehavior: LeaveMap` works fine as used in #18098.
The issue having to do with targeting makes sense. Aircraft do successfully reload and attack again on occasion, but will eventually stop if given enough time.

**Comment by cjshmyr** [comment]
> Something else in the aircraft code has regressed and so it remains stuck in one of the previous activities.

Brain dump time:

The Lua function works fine at feb5880, but on bleed it doesn't reach our [queued activity](https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/maps/allies-04/allies04-AI.lua#L266) if the target doesn't die.

Since then, we've had reworks to `FlyAttack` (d0f4414). With the new behavior in `FlyAttack`, and our aircraft having `AbortOnResupply: true`, it will cancel the next activity after its attack run. Hence, it's cancelling the Lua-queued activity.

`AbortOnResupply: false` means the aircraft will continue attacking and re-arming indefinitely until the target is dead. Changing to `false` works, but has an undesired side effect of player manned aircraft behaving this way (unless we go so far as defining different aircraft for AI... which sounds ugly).

I think this is new behavior we should follow rather than a bug, but could be corrected on that.

> Did we fix aircraft not triggering the idle trigger?
> The idle trigger never fires when the aircraft is flying in circles

I didn't see any recent or relevant changes in `FlyIdle`'s history. I was successful with testing `Trigger.OnIdle` at feb5880 and current bleed, with the following triggers for idle: 
- Aircraft has a `FlyIdle` trait ticking (plane flying in circle, or a still helicopter)
- Aircraft has finished re-arming

And:
- Does not trigger with `IdleBehavior: LeaveMap`
- `aircraft.IsIdle` will never be true since when circling, since `FlyIdle` is the current activity

My current suggestion is to use an idle trigger, but I'm not trying to force it. It has a down-side of needing to track persistent targets in a global table when triggered.

**Comment by pchote** [comment]
`FlyIdle` implements special logic to tick `INotifyIdle` (and therefore the idle trigger) while it is active. The fact that `.IsIdle` returns false in that case is IMO a bug.

**Comment by abcdefg30** [comment]
Apparently the trigger not getting fired was fixed by the manual `TickIdle` calls. I did the suggested fix for `IsIdle` in #13021 which was eventually reverted because of regressions.
